drivers/ThinQ_Washer.groovy

```diff
@@ -171,25 +157,39 @@ def processStateData(data) {
         }
     }
 
+    sendEvent(name: "runTime", value: runTime)
+    sendEvent(name: "runTimeDisplay", value: checkValue(data,'Remain_Time_H') ? "${data["Remain_Time_H"]}:${data["Remain_Time_M"]}" : "${data["Remain_Time_M"]}")
+    sendEvent(name: "remainingTime", value: remainingTime)
+    sendEvent(name: "remainingTimeDisplay", value: checkValue(data,'Initial_Time_H') ? "${data["Initial_Time_H"]}:${data["Initial_Time_M"]}" : "${data["Initial_Time_M"]}")
+    sendEvent(name: "delayTime", value: delayTime)
+    sendEvent(name: "delayTimeDisplay", value: checkValue(data,'Reserve_Time_H') ? "${data["Reserve_Time_H"]}:${data["Reserve_Time_M"]}" : "${data["Reserve_Time_M"]}")
 
-    if (data?.containsKey('Error') ) {
+    if (checkValue(data,'Error')) {
       sendEvent(name: "error", value: data["Error"].toLowerCase())
     }
 
-    if (data["APCourse"] != null && data["APCourse"] != "")
+    if (checkValue(data,'APCourse'))
         sendEvent(name: "course", value: data["APCourse"] != 0 ? data["APCourse"]?.toLowerCase() : "none")
-    if (data["SmartCourse"] != null && data["SmartCourse"] != "")
+    if (checkValue(data,'SmartCourse'))
         sendEvent(name: "smartCourse", value: data["SmartCourse"] != 0 ? data["SmartCourse"]?.toLowerCase() : "none")
-    if (soilLevel != null && soilLevel != "")
-        sendEvent(name: "soilLevel", value: parent.cleanEnumValue(soilLevel, "@WM_MX_OPTION_SOIL_"))
-    if (spinSpeed != null && spinSpeed != "")
-        sendEvent(name: "spinSpeed", value: parent.cleanEnumValue(spinSpeed, "@WM_MX_OPTION_SPIN_"))
-    if (temperatureLevel != null && temperatureLevel != "")
-        sendEvent(name: "temperatureLevel", value: parent.cleanEnumValue(temperatureLevel, "@WM_MX_OPTION_TEMP_"))
-    if (doorLock != null && doorLock != "")
-        sendEvent(name: "doorLock", value: parent.cleanEnumValue(doorLock, "@DOOR_LOCK_"))
+    if (checkValue(data,'Soil'))
+        sendEvent(name: "soilLevel", value: parent.cleanEnumValue(data["Soil"], "@WM_MX_OPTION_SOIL_"))
+    if (checkValue(data,'SpinSpeed'))
+        sendEvent(name: "spinSpeed", value: parent.cleanEnumValue(data["SpinSpeed"], "@WM_MX_OPTION_SPIN_"))
+    if (checkValue(data,'TempControl'))
+        sendEvent(name: "temperatureLevel", value: parent.cleanEnumValue(data["TempControl"], "@WM_MX_OPTION_TEMP_"))
+    if (checkValue(data,'doorLock'))
+        sendEvent(name: "doorLock", value: parent.cleanEnumValue(data["doorLock"], "@DOOR_LOCK_"))
 }
 
+// check is map has an actuale value
+private checkValue(data, String k) {
```

Comment:
I suspect we'll need this for all the drivers, so I'm going to move this to the app and we can call parent.checkValue, just to eliminate duplicate code.